### PR TITLE
Change processes of the General RNA-seq pipeline to support per-lane QC

### PR DESCRIPTION
**Pull request by OtonicarJan**
## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/expression/featureCounts_beta.py [hunk 1]
```diff
@@ -0,0 +1,980 @@
...
+    if len(filter_col) > 1:
+        exp["sum_count"] = exp[filter_col].sum(axis=1)
+        exp = exp.astype({"sum_count": int})
+
+        return exp[["sum_count"]].to_csv(
+            outfile,
+            index_label="FEATURE_ID",
+            header=["EXPRESSION"],
+            sep="\t",
+        )
```
Comment (AnzeLovse) [inline resolwe_bio/processes/expression/featureCounts_beta.py:51]: The insides of if statment could just sum counts and specify the return column. Then the return can be the same in both cases.
```
    if len(filter_col) > 1:
        return_column = "sum_count"
        exp[return_column] = exp[filter_col].sum(axis=1)
    else:
        return_column = sample_name


    exp = exp.astype({return_column: int})
    return exp[[return_column]].to_csv(
            outfile,
            index_label="FEATURE_ID",
            header=["EXPRESSION"],
            sep="\t",
    )
```

File: resolwe_bio/processes/expression/featureCounts_beta.py [hunk 2]
```diff
@@ -0,0 +1,980 @@
...
+
+        if inputs.exon_exon_junctions.junc_counts:
+            args.append("-J")
+
+        if inputs.exon_exon_junctions.junc_counts and inputs.exon_exon_junctions.genome:
+            args.extend(["-G", inputs.exon_exon_junctions.genome.path])
+
+        if inputs.general.by_read_group:
+            # Check if @RG is in header of the BAM file
+            (Cmd["samtools"]["view", "-H"][bam_file] > "header.sam")()
```
Comment (AnzeLovse) [inline resolwe_bio/processes/expression/featureCounts_beta.py:851]: `-o` option could be used to avoid piping and to improve error handling. If the output is a `.txt` file the check can be done in Python.

File: resolwe_bio/processes/reads_processing/bbduk_beta.py [hunk 3]
```diff
@@ -0,0 +1,1360 @@
...
+            f"xmin={inputs.header_parsing.x_min}",
+            f"ymin={inputs.header_parsing.y_min}",
+            f"xmax={inputs.header_parsing.x_max}",
+            f"ymax={inputs.header_parsing.y_max}",
+            f"entropy={inputs.complexity.entropy}",
+            f"entropywindow={inputs.complexity.entropy_window}",
+            f"entropyk={inputs.complexity.entropy_k}",
+            f"minbasefrequency={inputs.complexity.min_base_frequency}",
+            f"entropymask={inputs.complexity.entropy_mask}",
+            f"-Xmx{int(0.85*(self.requirements.resources.memory/num_of_lanes))}m",
```
Comment (AnzeLovse) [inline resolwe_bio/processes/reads_processing/bbduk_beta.py:601]: Let's make sure we thoroughly test resource handling on real data.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 4]
```diff
@@ -310,6 +310,182 @@ def filter_star_report(line):
...
+            reads = self.prepare_reads(
+                [str(input_folder / "hs_single bbduk_star_htseq_reads_single.fastq.gz")]
+            )
+            single_lanes = self.prepare_reads(
+                [
+                    str(
+                        input_folder
+                        / "hs_single bbduk_star_htseq_reads_single.fastq.gz"
+                    ),
+                    str("hs sim_reads_single.fastq.gz"),
```
Comment (AnzeLovse) [inline resolwe_bio/tests/processes/test_alignment.py:328]: No need for `str`.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 5]
```diff
@@ -310,6 +310,182 @@ def filter_star_report(line):
...
+                        input_folder / "hs_paired_R2 workflow_bbduk_star_htseq.fastq.gz"
+                    )
+                ],
+            )
+            paired_lanes = self.prepare_paired_reads(
+                mate1=[
+                    str(
+                        input_folder / "hs_paired_R1 workflow_bbduk_star_htseq.fastq.gz"
+                    ),
+                    str("hs sim_reads1.fastq.gz"),
```
Comment (AnzeLovse) [inline resolwe_bio/tests/processes/test_alignment.py:348]: Same as above.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 6]
```diff
@@ -310,6 +310,182 @@ def filter_star_report(line):
...
+                ],
+                mate2=[
+                    str(
+                        input_folder / "hs_paired_R2 workflow_bbduk_star_htseq.fastq.gz"
+                    ),
+                    str("hs sim_reads2.fastq.gz"),
```
Comment (AnzeLovse) [inline resolwe_bio/tests/processes/test_alignment.py:354]: Same as above.

File: resolwe_bio/tests/processes/test_expression.py [hunk 7]
```diff
@@ -494,6 +494,197 @@ def test_feature_counts(self):
...
+        )
+
+        expression_lanes = self.run_process(
+            "feature_counts-beta",
+            {
+                "aligned_reads": paired_lanes.id,
+                "annotation": annotation.id,
+            },
+        )
+        self.assertFile(
```
Comment (AnzeLovse) [inline resolwe_bio/tests/processes/test_expression.py:681]: If the process modified the `feature_counts_output` the tests would have to cover that as well.

File: resolwe_bio/processes/reads_processing/bbduk_beta.py [hunk 8]
```diff
@@ -0,0 +1,1360 @@
...
+        num_of_lanes = len(inputs.reads.output.fastq)
+        input_reads = []
+        for lane in range(num_of_lanes):
+            reads_basename = Path(inputs.reads.output.fastq[lane].path)
+            shutil.copy(reads_basename, Path.cwd())
+
+            new_path = Path(reads_basename.name.replace(" ", "_"))
+            Path(reads_basename.name).rename(new_path)
+            assert new_path.name.endswith(".fastq.gz")
+            name = new_path.name[:-9]
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/reads_processing/bbduk_beta.py:523]: Files have to be renamed, since bbduk does not handle spaces in the file names. Should we rename files in the previous step (upload of FASTQs, SRA)?

File: resolwe_bio/processes/reads_processing/bbduk_beta.py [hunk 9]
```diff
@@ -0,0 +1,1360 @@
...
+
+
+def prepare_fastqc(fastqgz, error):
+    """Prepare FastQC data for output."""
+    fastqc = []
+    fastqc_url = []
+    for fq in fastqgz:
+        reads_name = Path(fq).name.replace(".fastq.gz", "")
+        report_dir = Path("fastqc") / Path(f"{reads_name}_fastqc")
+        if not report_dir.is_dir():
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/reads_processing/bbduk_beta.py]: This is a check in Upload fastq reads processes, however it is not needed here, since the step of preparing fastqc outputs is a bit different.

File: resolwe_bio/processes/support_processors/multiqc.py [hunk 10]
```diff
@@ -499,9 +521,13 @@ def run(self, inputs, outputs):
 
             elif d.process.type == "data:expression:featurecounts:":
                 name = os.path.basename(d.output.counts_summary.path)
-                create_symlink(
-                    d.output.counts_summary.path, os.path.join(sample_dir, name)
+                sum_featurecounts_columns(
+                    summary_file=d.output.counts_summary.path, out_file=f"summed_{name}"
```
Comment (jkokosar) [inline resolwe_bio/processes/support_processors/multiqc.py:525]: Is the out file `f"summed_{name}"` placed in the sample dir?

Comment (OtonicarJan) [reply resolwe_bio/processes/support_processors/multiqc.py]: Fixed in the latest push.

File: resolwe_bio/processes/support_processors/multiqc.py [hunk 11]
```diff
@@ -144,6 +144,28 @@ def create_lib_strand_table(samples, reports):
...
+        dtype={
+            "Status": str,
+        },
+    )
+    if len(exp.columns) > 1:
+        sum_column = exp.columns[0].split(":")[0]
+        exp[sum_column] = exp.sum(axis=1)
+
+        exp = exp.astype({sum_column: int})
+        return exp[[sum_column]].to_csv(
```
Comment (AnzeLovse) [inline resolwe_bio/processes/support_processors/multiqc.py:162]: No need to return as `to_csv` returns None.

File: resolwe_bio/processes/reads_processing/bbduk_beta.py [hunk 12]
```diff
@@ -0,0 +1,1380 @@
...
+            f"entropywindow={inputs.complexity.entropy_window}",
+            f"entropyk={inputs.complexity.entropy_k}",
+            f"minbasefrequency={inputs.complexity.min_base_frequency}",
+            f"entropymask={inputs.complexity.entropy_mask}",
+            f"-Xmx{int(0.85*(self.requirements.resources.memory/num_of_lanes))}m",
+        ]
+
+        if self.requirements.resources.cores >= num_of_lanes:
+            args.append(
+                f"threads={int(self.requirements.resources.cores//num_of_lanes)}"
```
Comment (AnzeLovse) [inline resolwe_bio/processes/reads_processing/bbduk_beta.py:646]: If there are more lanes than cores this would be 0. Raising an informative error in such case should suffice.

**Review by jkokosar — APPROVED** [review]

**Review by AnzeLovse — APPROVED** [review]
